### DISCUSSION TITLE
Metrices : Value Recorder and Observer not being exported to Otel Collector.

### DISCUSSION THREAD

**Discussion by RSoodAMAT** [open-telemetry/opentelemetry-java#3139] [Q&A]
ValueRecorder and Observer not working  in latest version of open telemetry . They are not getting collected in Prometheus , with latest version.  Only Counter ones are being collected and shown at Prometheus. They were working fine with previous versions of open telemetry.

**Comment by jkwatson** [comment]
Can you provide an example that isn't working? When I try out our prometheus example, I do see the metrics being published on the configured port.

**Reply by RSoodAMAT** [reply]
java -javaagent:opentelemetry-javaagent-all.jar -Dotel.exporter.otlp.endpoint="http://localhost:4317" -jar -Dotel.resource.attributes=service.name=tc-job  TCapp.jar

static final Meter meter = GlobalMetricsProvider.getMeter("Train Center", "1");
// Observer example 
class Status{
private double avgDownloadSpeed; // corresponding getter and setter
}

private Status status;
meter
        .doubleValueObserverBuilder("avg.download.speed.per.wafer")
        .setDescription("Average Download speed per wafer")
        .setUnit("mb/sec")
        .setUpdater(
            doubleResult ->
                doubleResult.observe(
                    status.getAvgDownloadSpeed(),
                    Labels.of("wafer", mapID, "tool", toolId, "recipe", recipe, "unit", "mb/sec")));

// Recorder example 
long timeTaken =  System.currentTimeMillis();
meter
        .longValueRecorderBuilder("total.time.taken.indexing.task.per.wafer")
        .setDescription("Total time taken by dl hub indexing task per wafer")
        .setUnit("ms")
        .build()
        .record(
            timeTaken, Labels.of("wafer", waferId, "tool", toolId, "recipe", recipe, "unit", "ms"));



//**Docker compose file**   - Bringing up otel collector , prometheus , jaeger using the otlp exmaple given in git . 
version: "2"
services:

  # Jaeger
  jaeger-all-in-one:
    image: jaegertracing/all-in-one:latest
    ports:
      - "16686:16686"
      - "14268"
      - "14250"

  # Zipkin
  zipkin-all-in-one:
    image: openzipkin/zipkin:latest
    ports:
      - "9411:9411"

  # Collector
  otel-collector:
    image: ${OTELCOL_IMG}
    command: ["--config=/etc/otel-collector-config-demo.yaml", "${OTELCOL_ARGS}"]
    volumes:
      - ./otel-collector-config-demo.yaml:/etc/otel-collector-config-demo.yaml
    ports:
      - "1888:1888"   # pprof extension
      - "8888:8888"   # Prometheus metrics exposed by the collector
      - "8889:8889"   # Prometheus exporter metrics
      - "13133:13133" # health_check extension
      - "55678"       # OpenCensus receiver
      - "55681:55679" # zpages extension
      - "55680:55680" # otlp receiver
      - "4317:4317" # otlp receiver
    depends_on:
      - jaeger-all-in-one
      - zipkin-all-in-one

  # Prometheus
  prometheus:
    container_name: prometheus
    image: prom/prometheus:latest
    volumes:
      - ./prometheus.yaml:/etc/prometheus/prometheus.yml
    ports:
      - "9090:9090"


//////////////////////////////////////
**otel-collector donfig-demo. yaml** 

receivers:
  otlp:
    protocols:
        grpc:

exporters:
  prometheus:
    endpoint: "0.0.0.0:8889"
    namespace: traincenter
    const_labels:
      label1: value1
  logging:

  zipkin:
    endpoint: "http://zipkin-all-in-one:9411/api/v2/spans"
    format: proto

  jaeger:
    endpoint: jaeger-all-in-one:14250
    insecure: true

processors:
  batch:

extensions:
  health_check:
  pprof:
    endpoint: :1888
  zpages:
    endpoint: :55679

service:
  extensions: [pprof, zpages, health_check]
  pipelines:
    traces:
      receivers: [otlp]
      processors: [batch]
      exporters: [logging, zipkin, jaeger]
    metrics:
      receivers: [otlp]
      processors: [batch]
      exporters: [logging,prometheus]


/////////////////////////////////////////
**prometheus.yaml** 

scrape_configs:
  - job_name: 'otel-collector'
    scrape_interval: 2s
    static_configs:
      - targets: ['otel-collector:8889']
      - targets: ['otel-collector:8888']

**Reply by jkwatson** [reply]
I believe this is an issue in the collector. When I turn on the logging exporter in the collector, I can see the metrics being properly received, but they aren't ending up exposed to prometheus, I agree.

**Reply by jkwatson** [reply]
Discussion in the CNCF `#otel-collector` channel leads me to believe that this is the collector issue: https://github.com/open-telemetry/opentelemetry-collector/issues/2661

**Reply by odeke-em** [reply]
Yup I can confirm @jkwatson's response that those metrics are sent to the collector but just that the collector's Prometheus exporters don't support exporting the Summary type and I diagnosed this in https://github.com/open-telemetry/opentelemetry-collector/issues/2661#issuecomment-813896918 and I have a PR for the Prometheus exporter support in https://github.com/open-telemetry/opentelemetry-collector/pull/2900

**Reply by RSoodAMAT** [reply]
In which release version can we expect the above fix.

**Reply by odeke-em** [reply]
Could you please try again with the latest OpenTelemetry collector? We've done a bunch of work lately and fixed the related issues.

**Reply by RSoodAMAT** [reply]
Post updating the collector and the open telemetry version to 1.1.0, still facing the issues with the metrices not being recorded at Prometheus level . Both recorder and observer not working. The metrices couldn't be seen at Prometheus. 
At logging level I could see below lines , but couldn't see the same at Prometheus : 


otel-collector_1     | 2021-04-16T06:49:24.826Z	INFO	loggingexporter/logging_exporter.go:327	TracesExporter	{"#spans": 38}
otel-collector_1     | 2021-04-16T06:49:28.193Z	INFO	loggingexporter/logging_exporter.go:375	MetricsExporter	{"#metrics": 19}
otel-collector_1     | 2021-04-16T06:49:29.830Z	INFO	loggingexporter/logging_exporter.go:327	TracesExporter	{"#spans": 50}
otel-collector_1     | 2021-04-16T06:49:34.834Z	INFO	loggingexporter/logging_exporter.go:327	TracesExporter	{"#spans": 38}
otel-collector_1     | 2021-04-16T06:49:39.838Z	INFO	loggingexporter/logging_exporter.go:327	TracesExporter	{"#spans": 42}
otel-collector_1     | 2021-04-16T06:49:44.842Z	INFO	loggingexporter/logging_exporter.go:327	TracesExporter	{"#spans": 38}
otel-collector_1     | 2021-04-16T06:49:49.846Z	INFO	loggingexporter/logging_exporter.go:327	TracesExporter	{"#spans": 159}
otel-collector_1     | 2021-04-16T06:50:28.041Z	INFO	loggingexporter/logging_exporter.go:375	MetricsExporter	{"#metrics": 21}
otel-collector_1     | 2021-04-16T06:51:28.089Z	INFO	loggingexporter/logging_exporter.go:375	MetricsExporter	{"#metrics": 17}

**Reply by RSoodAMAT** [reply]
Any updates on the same, as it is not working.

**Reply by jkwatson** [reply]
@odeke-em  any updates? Also, can this discussion be moved to the collector repository?

**Reply by RSoodAMAT** [reply]
Ok thanks
